cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -445,6 +446,24 @@ public Map<ApplicationId, ApplicationMeta> getApplicationsForAppIds(Collection<A
                this::decodeRow));
   }
 
+  /**
+   * Update a map of applications with provided application Source Control Metadata.
+   * @param sourceControlMap the map for the applications' SourceControlMeta synced with linked repository.
+   */
+  public void setAppSourceControlMetas(Map<ApplicationId, SourceControlMeta> sourceControlMap) throws IOException {
+    StructuredTable appSpecTable = getApplicationSpecificationTable();
+    for (Map.Entry<ApplicationId, SourceControlMeta> entry : sourceControlMap.entrySet()) {
+      appSpecTable.upsert(getSourceControlMetaFields(entry.getKey(), entry.getValue()));
```

Comment:
do we need to do any checks here? For example, what if there are 2 pushes happening at the same time?

Reply:
Consider the case where A and B both do a push. A's commit and push happens before B's commit and push, but B gets to the DB update before A. 

Are we storing the commit time or anything like that? If so we can compare the time to whatever is in the DB and skip writing if we already have something newer there.

I think you already need to do a lookup anyway, as the current implementation is also open to races when somebody deletes the app at the same time. They could delete the app while the push is happening and remove all the rows, but then this goes and adds it back.

Reply:
@albertshau I added the check for application existence in case the app is deleted in push. I don't think it's a race for the case `where A and B both do a push. A's commit and push happens before B's commit and push, but B gets to the DB update before A`: since they're writing on the same latest version, the fileHash to write is also the same. Actually the current code throws exception in the case when the configs have no update than the one in git.

Reply:
We might need to better define the response for the case that the application to write has no changes than the one in git